Added support for PSWAP and ISWAP gates in runtime for Braket plugin

paul0403 opened the pull request:
… (#395)

Specifically,
- Updated frontend/catalyst/pennylane_extensions.py to include "PSWAP" and "ISWAP" gates
- Updated runtime/include/RuntimeCAPI.h and runtime/lib/capi/RuntimeCAPI.cpp to interface the pswap and iswap gates from Braket plugin, as the Braket QuantumDevice's NamedOperation.
- Updated runtime/lib/backend/openqasm/OpenQasmBuilder.hpp to include the two gates so that the backend can look them up from OpenQasm.
  - The two lines are already there in OpenQasmBulider.hpp; I merely added two comments. Maybe it was inadvertently left in when making the assignment for me?

This is Haochen (Paul) Wang's technical interview.

### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new functions and code must be clearly commented and documented.

- [x] Ensure that code is properly formatted by running `make format`.
      The latest version of black and `clang-format-14` are used in CI/CD to check formatting.

- [x] All new features must include a unit test.
      Integration and frontend tests should be added to [`frontend/test`](../frontend/test/),
      Quantum dialect and MLIR tests should be added to [`mlir/test`](../mlir/test/), and
      Runtime tests should be added to [`runtime/tests`](../runtime/tests/).

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** We would like to add support for the PennyLane PSWAP and ISWAP gates to our compilation architecture. Originally, no Catalyst device supported them natively, so they weren't considered, but the Braket plugin which was supported at a later point in time does come with support for them. 

**Description of the Change:** Added the PSWAP and ISWAP gates to refer to the Amazon Braket pswap and iswap gates in the backend. Specifically,
- Updated frontend/catalyst/pennylane_extensions.py to include "PSWAP" and "ISWAP" gates;
- Updated runtime/include/RuntimeCAPI.h and runtime/lib/capi/RuntimeCAPI.cpp to interface the pswap and iswap gates from Braket plugin, as the Braket QuantumDevice's NamedOperation;
- Updated runtime/lib/backend/openqasm/OpenQasmBuilder.hpp to include the two gates so that the backend can look them up from OpenQasm;
- Updated runtime/lib/backend/common/Utils.hpp to include the PSWAP and ISWAP gates.

**Benefits:** A circuit like 
```python
import pennylane as qml
from catalyst import *

dev = qml.device("braket.local.qubit", wires=2, shots=100)

@qjit
@qml.qnode(dev)
def f(x: float):
    qml.Hadamard(0)
    qml.PSWAP(x, wires=[0, 1])
    qml.ISWAP(wires=[1, 0])
    return qml.probs()
```
would no longer decompose the PSWAP and ISWAP gates into braket backend's SWAPs and CNOTs and Hadamards. Instead it would just call braket backend's PSWAP and ISWAP:

```python
{ lambda ; a:f64[]. let
    b:f64[4] = func[
      call_jaxpr={ lambda ; c:f64[]. let
           = qdevice[
            rtd_kwargs={'shots': 100, 'device_type': 'braket.local.qubit', 'backend': 'braket_sv'}
            rtd_lib=/home/paul/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_openqasm.so
            rtd_name=OpenQasmDevice
          ] 
          d:AbstractQreg() = qalloc 2
          e:AbstractQbit() = qextract d 0
          f:AbstractQbit() = qinst[op=Hadamard qubits_len=1] e
          g:AbstractQbit() = qextract d 1
          h:AbstractQbit() i:AbstractQbit() = qinst[op=PSWAP qubits_len=2] f g c
          j:AbstractQbit() k:AbstractQbit() = qinst[op=ISWAP qubits_len=2] i h
          l:AbstractObs(num_qubits=2,primitive=compbasis) = compbasis k j
          m:f64[4] = probs[shape=(4,)] l
          n:AbstractQreg() = qinsert d 0 k
          o:AbstractQreg() = qinsert n 1 j
           = qdealloc o
        in (m,) }
      fn=<QNode: wires=2, device='braket.local.qubit', interface='auto', diff_method='best'>
    ] a
  in (b,) }
```

**Possible Drawbacks:** To register ISWAP and PSWAP as supported for OpenQasm, they needed to be added to the utilities ` enum class Ligntning::SimulatorGate` and `std::array Ligntning::simulator_gate_info` in Utils.hpp. Because all the three NamedOperation() for the three devices share the same `Lightning::lookup_gates()` from common/Utils.hpp, and PSWAP and ISWAP are only supported on OpenQasm, the NamedOperation() for Lightning and LightningKokkos are modified so that they do not try to fetch any unsupported ISWAP and PSWAP.

This issue fundamentally arises from the fact that the three devices share the same `lookup_gates()` from common/Utils.hpp in their respective `LightningSimulator/LightningKokkosSimulator/OpenQasmDevice::NamedOperation()`. 

As a workaround, the ISWAP and PSWAP NamedOperation() for Lightning and LightningKokkos are right now hard-coded by their matrices explicitly. I suggest the removal of a shared utility array and switch to another kind of gate support in C++ simulator runtime to avoid similar issues in the future.

**Related GitHub Issues:** #395

josh146 commented:
Thanks @paul0403! I've opened a separate PR to fix the docs build in #460

maliasadi commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -227,6 +227,8 @@ class QJITDevice(qml.QubitDevice):
         "MultiRZ",
         "QubitUnitary",
+        "ISWAP",
+        "PSWAP",
```
Let's ensure these gates are now supported from the frontend. You can check the support for lightning simulators in `test_operations.py` and for openqasm devices in `test_braket_local_devices.py` and `test_braket_aws_devices.py`.

dime10 replied:
I would also add that we should write a test that verifies the gates are not decomposed :)

dime10 replied:
Thanks for adding some tests! The decomposition aspect happens in the frontend actually, so it would be great to have a test there that checks the gates are not decomposed (rather than in the runtime). 

In this case, you could for example check that `ISWAP` and `PSWAP` appear in the jaxpr of a test case. 

Tests that verify something about our generated IR are well suited for lit/FileCheck tests. Have a look at our existing lit tests to see how to write one: https://github.com/PennyLaneAI/catalyst/tree/main/frontend/test/lit

maliasadi reviewed (commented):
Thanks @paul0403! :slightly_smiling_face:

maliasadi commented on runtime/lib/capi/RuntimeCAPI.cpp:
```diff
@@ -580,6 +580,20 @@ void __quantum__qis__MultiRZ(double theta, bool adjoint, int64_t numQubits, ...)
+void __quantum__qis__ISWAP(QUBIT *wire0, QUBIT *wire1, bool adjoint)
+{
+    Catalyst::Runtime::getQuantumDevicePtr()->NamedOperation(
+        "ISWAP", {}, {reinterpret_cast<QubitIdType>(wire0), reinterpret_cast<QubitIdType>(wire1)},
```
Why do we need these `reinterpret_cast`? Is there any way that we can rewrite this code without casting?

dime10 replied:
Isn't that what we are doing for other gates as well?

maliasadi replied:
just wanted to know @paul0403's thoughts on this :)

dime10 replied:
Nice insight!

maliasadi replied:
Awesome! :guitar:

maliasadi commented on runtime/lib/backend/common/Utils.hpp:
```diff
@@ -198,6 +200,8 @@ constexpr std::array simulator_gate_info = {
     GateInfoTupleT{SimulatorGate::CZ, "CZ", 2, 0},
     GateInfoTupleT{SimulatorGate::SWAP, "SWAP", 2, 0},
+    GateInfoTupleT{SimulatorGate::ISWAP, "ISWAP", 2, 0},
+    GateInfoTupleT{SimulatorGate::PSWAP, "PSWAP", 2, 1},
```
What are your thoughts about removing these utility arrays? What would you do if you wanted to re-implement the gate support in the runtime from the beginning?

maliasadi commented on runtime/include/RuntimeCAPI.h:
```diff
@@ -77,6 +77,8 @@ void __quantum__qis__CRot(double, double, double, QUBIT *, QUBIT *, bool);
 void __quantum__qis__CSWAP(QUBIT *, QUBIT *, QUBIT *, bool);
 void __quantum__qis__Toffoli(QUBIT *, QUBIT *, QUBIT *, bool);
 void __quantum__qis__MultiRZ(double, bool /*adjoint*/, int64_t, /*qubits*/...);
+void __quantum__qis__ISWAP(QUBIT *, QUBIT *, bool /*adjoint*/);
```
How would the compiler know about the existence of these new gates?

maliasadi replied:
Thanks @paul0403! :100:

paul0403 commented:
Sorry I have classes until 9pm today, I'll look into these comments tomorrow!

paul0403 commented:
@josh146 I rebased this feature branch so that the upstream changes from main is included. Please run CI checks. Thanks. 

@maliasadi @dime10 Technically my assessment is due tomorrow (Friday Jan.26) so I would appreciate some comments and the finishing up of this work. Thanks.

Also there seems to be conflicts in changelog.md that I can't resolve from my side because I don't have write access.

dime10 commented:
Hi @paul0403, thanks for providing an update :) Would you mind also responding to open comments describing what you've addressed or providing additional information?

> Also there seems to be conflicts in changelog.md that I can't resolve from my side because I don't have write access.

I'm guessing you mean you can't resolve the conflict through the GitHub UI, but you should be able to resolve the conflict on your own branch/fork by merging with the latest version of main.

dime10 commented:
Can you verify that your comments are not marked "pending" or something similar? This can sometimes happen when a review is started, which blocks comments from being submitted until the review is submitted.

dime10 commented:
Oh my, yeah this is really unfortunate design on GitHub's side (has happened to me plenty of times as well) :/

You can submit them by going to the files tab and clicking "submit review". The type of review can just be "comment".

josh146 commented:
Hey @paul0403! it sounds like you have submitted your replies as a review. In which case, you will need to submit the review for it to become visible:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/0b44ac62-dcc8-4701-8657-830cfee2094f)

paul0403 reviewed (commented):
Seems like this is how I submit comments

dime10 commented:
Thanks @paul0403, I had a look your responses and you have some great insights! I'll let @maliasadi respond/resolve them since he opened them.

In order for us to merge the PR (not a requirement for the deadline), I would like us to find a different way of resolving the utility array issue that doesn't rely on adding "manual support" for these gates to the lightning backends, since the gates are not supported in the upstream device and we don't want to maintain such patches. It's great that you found a way to make it work and get the tests to pass though!

maliasadi commented on doc/changelog.md:
```diff
@@ -2,6 +2,62 @@
+This issue fundamentally arises from the fact that the three devices share the same `lookup_gates()` from common/Utils.hpp in their respective `LightningSimulator/LightningKokkosSimulator/OpenQasmDevice::NamedOperation()`. 
+
+As a workaround, the ISWAP and PSWAP NamedOperation() for Lightning and LightningKokkos are right now hard-coded by their matrices explicitly. I suggest the removal of a shared utility array and switch to another kind of gate support in C++ simulator runtime to avoid similar issues in the future.
+  [(#458)](https://github.com/PennyLaneAI/catalyst/pull/458)
```
Don't forget to add your name to the list of contributors :slightly_smiling_face:

maliasadi commented on runtime/lib/capi/RuntimeCAPI.cpp:
```diff
@@ -580,6 +580,20 @@ void __quantum__qis__MultiRZ(double theta, bool adjoint, int64_t numQubits, ...)
+void __quantum__qis__PSWAP(double phi, QUBIT *wire0, QUBIT *wire1, bool adjoint)
+{
+    Catalyst::Runtime::getQuantumDevicePtr()->NamedOperation(
+        "PSWAP", {phi},
```
Hey @paul0403! Thanks for looking into this! Except classifying quantum gates, I can think of two feasible options: 1) Each device can have it's own list of supported gates. 2) Rely on backend devices APIs to check supported gates.

We have followed the second approach for all of our devices. But unfortunately, Lightning simulators are not able to raise meaningful error messages anymore! I will ensure that this will be fixed in the near future. For the scope of this PR, I think it should be enough to add a `RT_FAIL_IF` to the beginning of `NamedOperation` to raise an error if users call this method with `ISWAP` or `PSWAP` for now. You can update your C++ test to check this patch using `REQUIRE_THROWS_WITH`. 

Note that these gates are decomposed in the frontend for Lightning simulators, so there should not be any issues for users to use these gates from the frontend.

maliasadi replied:
@paul0403 let's remove this C++ test and all the `PSWAP` and `ISWAP` patches to `LightningSimulator` and `LightningKokkosSimulator` in this PR. The issue is related to the Lightning C++ API in https://github.com/PennyLaneAI/pennylane-lightning that we will provide a fix for it later. So, just ensure we have C++ test for the OpenQasm device. Thank you for digging into this issue! :rocket:

paul0403 replied:
@maliasadi Will do. Is there a way to indicate in `runtime/tests/Test_LightningGateSet.cpp` that a test should only be run on openqasm device?

maliasadi replied:
`Test_LightningGateSet.cpp` is a test suite for Lightning simulators tests. you can add your openqasm unit tests to `Test_OpenQasmDevice.cpp` and `Test_OpenQasmBuilder` to test on OpenQasm devices.

Based on the latest version of your work, you only need to add something like the following test to `Test_OpenQasmDevice.cpp`:

``` c++
TEST_CASE("Test PSWAP and ISWAP with BuilderType::Braket", "[openqasm]")
{
    constexpr size_t shots{1000};
    std::unique_ptr<OpenQasmDevice> device = std::make_unique<OpenQasmDevice>(
        "{device_type : braket.local.qubit, backend : default, shots : 1000}");

    constexpr size_t n{2};
    auto wires = device->AllocateQubits(n);

    device->NamedOperation("Hadamard", {}, {wires[0]}, false);
    device->NamedOperation("ISWAP", {}, {wires[0], wires[1]}, false);
    device->NamedOperation("PSWAP", {0.34}, {wires[0], wires[1]}, false);

    auto obs = device->Observable(ObsId::PauliZ, {}, std::vector<QubitIdType>{1});
    auto expval = device->Expval(obs);
    CHECK(expval == Approx(1).margin(1e-5));
}
```

paul0403 commented:
Hi all, thanks for the comments. I'll work on them later as I have class until the end of day today. I expect all of them to be resolved either tonight or tomorrow.

paul0403 commented:
@maliasadi @dime10 I have worked through all the comments and added the suggested frontend and openqasm tests. `make test-runtime ENABLE_OPENQASM=ON` and `make test-frontend ENABLE_OPENQASM=ON` both pass now. @josh146 can run CI checks and hopefully we'll be good (except that one annoying sphinx version 4.5<5 error which I have no idea what I could do). 

I left in the PSWAP and ISWAP capi support & tests for Lightning and LightningKokkos as comments for now. You can decide whether we should keep them in as comments for straight up delete them. 

If CI checks are good I'll rebase to keep up with main branch and squash&cleanup all commits.

maliasadi commented on frontend/test/lit/test_multi_qubit_gates.py:
```diff
@@ -62,3 +62,17 @@ def circuit():
+@qjit(target="mlir")
+@qml.qnode(qml.device("braket.local.qubit", wires=2, shots=100))
+def circuit(x: float):
+    # CHECK: {{%.+}} = quantum.custom "ISWAP"() {{.+}} : !quantum.bit, !quantum.bit
```
@paul0403 could you address these codefactor suggestions?

maliasadi replied:
https://github.com/PennyLaneAI/catalyst/pull/458/checks?check_run_id=20880970204

paul0403 replied:
This added function is in accordance with the style of the rest of the file frontend/test/lit/test_multi_qubit_gates.py. They are all named "circuit" and don't have docstrings.

maliasadi replied:
we should likely fix them all sometime :t-rex:  don't worry about these codefactor CI suggestions there! This is not a strong merge requirement

maliasadi commented on runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp:
```diff
@@ -133,6 +136,28 @@ void LightningSimulator::NamedOperation(const std::string &name, const std::vect
+            };
+            LightningSimulator::MatrixOperation(matrix, wires, inverse);
+        }
+    */
```
You can remove these commented lines. This issue will be resolved upstream in the pennylane-lightning repository.

paul0403 replied:
done

maliasadi commented on runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp:
```diff
@@ -115,6 +115,9 @@ auto LightningSimulator::One() const -> Result
 void LightningSimulator::NamedOperation(const std::string &name, const std::vector<double> &params,
                                         const std::vector<QubitIdType> &wires, bool inverse)
 {
+    // Although ISWAP and PSWAP exist in Lightning::simulator_gate_info, they are only supported on
```
The list of `simulator_gate_info` is there to check the number of parameters and wires for each requested gate on  simulators. This array is not specific to Lightning simulators. Users should be able to pass any gate name and backend devices should provide neat error messages for unsupported gates. This is currently missing for the Lightning simulator, but I ensure this  will be added upstream.

maliasadi replied:
You can remove line 118-120.

paul0403 replied:
done

maliasadi commented on runtime/lib/backend/lightning/lightning_kokkos/LightningKokkosSimulator.cpp:
```diff
@@ -129,6 +132,28 @@ void LightningKokkosSimulator::NamedOperation(const std::string &name,
+            };
+            LightningKokkosSimulator::MatrixOperation(matrix, wires, inverse);
+        }
+    */
```
Same as above. You can remove this commented workaround.

maliasadi commented on runtime/tests/Test_LightningGateSet.cpp:
```diff
@@ -478,6 +478,63 @@ TEMPLATE_LIST_TEST_CASE("CSWAP test", "[GateSet]", SimTypes)
+    CHECK(state[2].imag() == Approx(-0.5).epsilon(1e-5));
+    CHECK(state.at(3) == std::complex<double>{0, 0});
+}
+*/
```
You can keep these tests and use `REQUIRE_THROWS_WITH` to check they are failed with some key-value exceptions. I'd be okay if you leave them as is too. We can revisit them after adding the patch upstream to the pennylane-lightning repository,

paul0403 replied:
I will keep them as is. I believe in some other test files (e.g. `test_operations.py`) there are commented tests waiting to be uncommented once the support arrives.

maliasadi reviewed (commented):
Nice work @paul0403! Please address codefactor suggestions and this should be good to go :shark:

paul0403 commented:
@maliasadi I have squashed commits and rebased to keep up with main. Please run CI: we should be all good. 

For the codefactor failure, the added unit test function is in accordance with the style of the rest of the file frontend/test/lit/test_multi_qubit_gates.py. They are all named "circuit" and don't have docstrings.

paul0403 commented:
@maliasadi  Alright the CIs have passed (except codefactor) and we're on track with main branch. I think we're good to go, unless there's other things you want to go over from the manual review. If it's all good I'll let Aaron know! 

It was fun going deep into your codebase. It was also surprising to learn what "pending" on github actually means...

maliasadi commented on doc/changelog.md:
```diff
@@ -3,6 +3,52 @@
 <h3>New features</h3>
 
 <h3>Improvements</h3>
+* Added the PSWAP and ISWAP gates to refer to the Amazon Braket pswap and iswap gates in the backend. Specifically, a circuit like
```
```suggestion
* Add native support for `qml.PSWAP` and `qml.ISWAP` gates on Amazon Braket devices. Specifically, a circuit like
```

maliasadi commented on doc/changelog.md:
````diff
@@ -3,6 +3,52 @@
+      return qml.probs()
+  ```
+
+would no longer decompose the PSWAP and ISWAP gates into braket backend's SWAPs and CNOTs and Hadamards. Instead it would just call braket backend's PSWAP and ISWAP:
````
```suggestion
would no longer decompose the `PSWAP` and `ISWAP` gates to `SWAP`s, `CNOT`s and `Hadamard`s. Instead it would just call Braket's native `PSWAP` and `ISWAP` gates at runtime.
```

maliasadi commented on doc/changelog.md:
````diff
@@ -3,6 +3,52 @@
+ 
+  ```py
+  import pennylane as qml
+  from catalyst import *
````
```suggestion
  from catalyst import qjit
```

maliasadi commented on doc/changelog.md:
````diff
@@ -3,6 +3,52 @@
+  ```
+
+would no longer decompose the PSWAP and ISWAP gates into braket backend's SWAPs and CNOTs and Hadamards. Instead it would just call braket backend's PSWAP and ISWAP:
+  
+  ```py
+  { lambda ; a:f64[]. let
+      b:f64[4] = func[
+        call_jaxpr={ lambda ; c:f64[]. let
+             = qdevice[
+              rtd_kwargs={'shots': 100, 'device_type': 'braket.local.qubit', 'backend': 'braket_sv'}
+              rtd_lib=/home/paul/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_openqasm.so
+              rtd_name=OpenQasmDevice
+            ] 
+            d:AbstractQreg() = qalloc 2
+            e:AbstractQbit() = qextract d 0
+            f:AbstractQbit() = qinst[op=Hadamard qubits_len=1] e
+            g:AbstractQbit() = qextract d 1
+            h:AbstractQbit() i:AbstractQbit() = qinst[op=PSWAP qubits_len=2] f g c
+            j:AbstractQbit() k:AbstractQbit() = qinst[op=ISWAP qubits_len=2] i h
+            l:AbstractObs(num_qubits=2,primitive=compbasis) = compbasis k j
+            m:f64[4] = probs[shape=(4,)] l
+            n:AbstractQreg() = qinsert d 0 k
+            o:AbstractQreg() = qinsert n 1 j
+             = qdealloc o
+          in (m,) }
+        fn=<QNode: wires=2, device='braket.local.qubit', interface='auto', diff_method='best'>
+      ] a
+    in (b,) }
+  ```
+
````
This can be removed from the changelog.

```suggestion
```

maliasadi commented on runtime/tests/Test_LightningGateSet.cpp:
```diff
@@ -478,6 +478,63 @@ TEMPLATE_LIST_TEST_CASE("CSWAP test", "[GateSet]", SimTypes)
     CHECK(state[5].imag() == Approx(0).epsilon(1e-5));
 }
 
+/*
```
```suggestion
// TODO: Uncomment these tests after `PSWAP` and `ISWAP` are natively supported by Lightning simulators.  
/*
```

maliasadi reviewed (approved):
:partying_face:

paul0403 commented:
@maliasadi Updated with the suggested changes and squashed into just one commit. We should be all good now as soon as CI gives the green light.

erick-xanadu commented:
Hi @paul0403! I just recently made some changes to the runtime. I think you will need to change the functions you recently defined as `__quantum__` to be `__catalyst__`.

paul0403 commented:
@erick-xanadu @maliasadi done, please run CI

paul0403 commented:
@maliasadi I think we're good to merge! If there's no other work, please merge and I'll let Aaron know that I have finished the technical assessment. Thanks for all the help!

maliasadi commented on doc/changelog.md:
```diff
@@ -3,6 +3,25 @@
 <h3>New features</h3>
 
 <h3>Improvements</h3>
+* Add native support for `qml.PSWAP` and `qml.ISWAP` gates on Amazon Braket devices. Specifically, a circuit like
```
There's one more MD rendering issue :)
```suggestion

* Add native support for `qml.PSWAP` and `qml.ISWAP` gates on Amazon Braket devices. Specifically, a circuit like
```

paul0403 replied:
@maliasadi  Updated

maliasadi commented:
Hey @paul0403, thanks for adding the support! Feel free to let Aaron know that you are done with the assessment.

paul0403 commented:
@maliasadi some other merge request (seems like #483) got approved in the past couple of minutes, giving a conflict in changelog.md. Just resolved it. Let's try one more time.

paul0403 commented:
![image](https://github.com/PennyLaneAI/catalyst/assets/79805239/72a98545-2a84-4bf3-bc70-0eebe945e048)

The failure seems to not come from my end @maliasadi

maliasadi commented:
@paul0403 fixed now!

paul0403 commented:
Nice! Please merge when you feel appropriate. @maliasadi

maliasadi commented:
:tada:
